src/ucs/datastruct/arbiter.c

```diff
@@ -49,6 +50,49 @@ void ucs_arbiter_group_push_elem_always(ucs_arbiter_group_t *group, ucs_arbiter_
     group->tail = elem;   /* Update group tail */
 }
 
+static void ucs_arbiter_group_head_replaced(ucs_arbiter_t *arbiter,
+                                            ucs_arbiter_elem_t *old_head,
+                                            ucs_arbiter_elem_t *new_head)
+{
+    ucs_assert(old_head->list.prev != NULL);
+    ucs_assert(old_head->list.next != NULL);
+    ucs_assert(old_head != new_head);
+
+    ucs_list_insert_replace(old_head->list.prev, old_head->list.next,
+                            &new_head->list);
+    if (arbiter->current == old_head) {
+        arbiter->current = new_head;
+    }
+}
+
+void ucs_arbiter_group_push_head_elem_always(ucs_arbiter_t *arbiter,
+                                             ucs_arbiter_group_t *group,
+                                             ucs_arbiter_elem_t *elem)
+{
+    ucs_arbiter_elem_t *tail = group->tail;
+    ucs_arbiter_elem_t *head;
+
+    elem->group = group;  /* Always point to group */
+
+    if (tail == NULL) {
+        elem->list.next = NULL;   /* Not scheduled yet */
+        elem->next      = elem;   /* Connect to itself */
+        group->tail     = elem;   /* Update group tail */
+        return;
+    }
+
+    head       = tail->next;
+    elem->next = head;        /* Point to first element */
+    tail->next = elem;        /* Point previous element to new one */
+
+    if (head->list.next != NULL) {
+        ucs_assert(arbiter != NULL);
+        ucs_arbiter_group_head_replaced(arbiter, head, elem);
+    } else {
+        elem->list.next = NULL; /* Mark the new head as un-scheduled */
```

Comment:
minor: seems you can move this init to the beginning of the func and this branch (and line 78) will not be needed